File: src/01_basics/basics_cpp/src/topic_listener.cpp
// TUTORIAL 1b — Topics: Subscriber (Listener)
//
// ros2 run basics_cpp topic_talker     (separate terminal)
// ros2 run basics_cpp topic_listener

#include <memory>
#include "rclcpp/rclcpp.hpp"
#include "std_msgs/msg/string.hpp"

class Listener : public rclcpp::Node
{
public:
  Listener() : Node("listener")
  {
    // === CHECKPOINT: topics ===
    sub_ = create_subscription<std_msgs::msg::String>(
      "chatter", 10,
      [this](const std_msgs::msg::String & msg) {
        RCLCPP_INFO(get_logger(), "Heard: \"%s\"", msg.data.c_str());
      });
    RCLCPP_INFO(get_logger(), "Listener started — waiting on /chatter");
    // === END CHECKPOINT: topics ===
  }

private:
  // === CHECKPOINT: topics ===
  rclcpp::Subscription<std_msgs::msg::String>::SharedPtr sub_;
  // === END CHECKPOINT: topics ===
};

int main(int argc, char * argv[])
{
  rclcpp::init(argc, argv);
  rclcpp::spin(std::make_shared<Listener>());
  rclcpp::shutdown();
}

File: src/01_basics/basics_py/basics_py/topic_listener.py
"""
TUTORIAL 1b — Topics: Subscriber (Listener)

Run alongside topic_talker.py.

    ros2 run basics topic_talker     (separate terminal)
    ros2 run basics topic_listener
"""

import rclpy
from rclpy.node import Node
from std_msgs.msg import String


class Listener(Node):
    def __init__(self):
        super().__init__('listener')

        # === CHECKPOINT: topics ===
        self.sub = self.create_subscription(String, 'chatter', self.on_message, 10)
        self.get_logger().info('Listener started — waiting on /chatter')
        # === END CHECKPOINT: topics ===

    # === CHECKPOINT: topics ===
    def on_message(self, msg: String):
        self.get_logger().info(f'Heard: "{msg.data}"')
    # === END CHECKPOINT: topics ===


def main(args=None):
    rclpy.init(args=args)
    node = Listener()
    rclpy.spin(node)
    node.destroy_node()
    rclpy.shutdown()
